python/target_selection/cartons/mwm_yso.py

```diff
@@ -0,0 +1,576 @@
+                                             TwoMassPSC, YSO_Clustering)
+
+# from . import BaseCarton
+from target_selection.cartons import BaseCarton
+
+
+# See catalog.py for the name of peewee model names corresponding
+# to postgres table names:
+# https://github.com/sdss/sdssdb/blob/master/python/sdssdb/peewee/sdss5db/catalogdb.py
+
+
+class MWM_YSO_S1_Carton(BaseCarton):
+    """2.5.1. YSOs - S1 (IR excess)
+    Shorthand name: mwm_yso_s1
+Simplified Description of selection criteria:
+selection of YSOs based on IR excess,
+with WISE colors W1-W2>0.25, W2-W3>0.5, W3-W4>1.5,
+closer than parallax>0.3, and brighter than H<13
+(should have ~21.5K sources)
+Wiki page: https://wiki.sdss.org/display/MWM/YSO+selection+function
+Additional source catalogs needed: Gaia, 2mass, allwise
+Additional cross-matching needed:
+Note: Using the Gaia xmatch somehow misses half the sources.
+Selection was done on the allwise catalog that had 2mass photometry,
+and then the resulting selection was crossmatched against against
+Gaia with 1" search radius.
+Return columns: Gaia id, 2mass id, allwise id, G, BP, RP,
+J, H, K, W1, W2, W3, W4,parallax
+cadence options for these targets
+(list all options,
+even though no single target will receive more than one):
+Pseudo SQL (optional):
+Implementation: h_m<13 and w1mpro-w2mpro>0.25 and
+ w2mpro-w3mpro>0.5 and w3mpro-w4mpro>1.5 and parallax>0.3
+    """
+    name = 'mwm_yso_s1'
+    category = 'science'
+    cadence = None
+    program = 'program_mwm_yso_s1'
+    mapper = 'MWM'
+
+# Implementation: h_m<13 and w1mpro-w2mpro>0.25 and w2mpro-w3mpro>0.5 and
+# w3mpro-w4mpro>1.5 and parallax>0.3
+
+    def build_query(self, version_id, query_region=None):
+        query = (CatalogToTIC_v8
+                 .select(CatalogToTIC_v8.catalogid)
+                 .join(TIC_v8, on=(CatalogToTIC_v8.target_id == TIC_v8.id))
+                 .join(Gaia_DR2, on=(TIC_v8.gaia_int == Gaia_DR2.source_id))
+                 .switch(TIC_v8)
+                 .join(TwoMassPSC, on=(TIC_v8.twomass_psc == TwoMassPSC.designation))
+                 .switch(TIC_v8)
+                 .join(AllWise, on=(TIC_v8.allwise == AllWise.designation))
+                 .where(CatalogToTIC_v8.version_id == version_id,
+                        CatalogToTIC_v8.best >> True,
+                        TwoMassPSC.h_m < 13,
+                        (AllWise.w1mpro - AllWise.w2mpro) > 0.25,
+                        (AllWise.w2mpro - AllWise.w3mpro) > 0.50,
+                        (AllWise.w3mpro - AllWise.w4mpro) > 1.50,
+                        Gaia_DR2.parallax > 0.3))
+
+# Gaia_DR2 pweewee model class corresponds to
+# table catalogdb.gaia_dr2_source.
+#
+# All values of TIC_v8.plx (for non-null entries) are not the same as
+# values of Gaia_DR2.parallax.
+# Hence, we cannot use TIC_v8.plx instead
+# of Gaia_DR2.parallax.
+
+        if query_region:
+            query = query.where(peewee.fn.q3c_radial_query(Catalog.ra,
```

Comment:
For this to work you need yo join with `Catalog` because `query` doesn't contain it anymore. Same for the other cartons.